File: nf-dream/scripts/translate_prescan.py
#!/usr/bin/env python3
"""nf-dream translate pre-scan.

For each filtered file, count lines containing Vietnamese diacritics. Buckets:
  none   — 0 VN lines
  light  — 1-2 VN lines (skip by default; likely proper noun)
  mixed  — 3+ VN lines but < 20% of total lines
  heavy  — VN lines >= 20% of total

Candidates: mixed + heavy. Output JSON for Claude session to pick translation targets.

Usage:
  translate_prescan.py <scope> [--project <slug>] [--json]
"""
from __future__ import annotations

import argparse
import json
import re
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _lib import iter_memory_files, parse_frontmatter, extract_frontmatter_dict  # noqa: E402

DIACRITICS = (
    "àáảãạăắằẳẵặâấầẩẫậ"
    "èéẻẽẹêếềểễệ"
    "ìíỉĩị"
    "òóỏõọôốồổỗộơớờởỡợ"
    "ùúủũụưứừửữự"
    "ỳýỷỹỵ"
    "đ"
    "ÀÁẢÃẠĂẮẰẲẴẶÂẤẦẨẪẬ"
    "ÈÉẺẼẸÊẾỀỂỄỆ"
    "ÌÍỈĨỊ"
    "ÒÓỎÕỌÔỐỒỔỖỘƠỚỜỞỠỢ"
    "ÙÚỦŨỤƯỨỪỬỮỰ"
    "ỲÝỶỸỴ"
    "Đ"
)
DIACRITIC_RE = re.compile(f"[{re.escape(DIACRITICS)}]")


def bucket_of(vn_lines: int, total_lines: int) -> str:
    if vn_lines == 0:
        return "none"
    if vn_lines <= 2:
        return "light"
    ratio = vn_lines / max(total_lines, 1)
    return "heavy" if ratio >= 0.20 else "mixed"


def scan_file(path: Path) -> dict:
    text = path.read_text()
    parsed = parse_frontmatter(text)
    body = parsed[1] if parsed else text
    lines = body.split("\n")
    vn_lines = sum(1 for ln in lines if DIACRITIC_RE.search(ln))
    bucket = bucket_of(vn_lines, len(lines))
    return {
        "file": path.name,
        "total_lines": len(lines),
        "vn_lines": vn_lines,
        "bucket": bucket,
        "candidate": bucket in ("mixed", "heavy"),
    }


def project_matches(fm_dict: dict, slug: str | None) -> bool:
    if slug is None:
        return True
    proj = fm_dict.get("metadata.project", "")
    return slug in proj or "generic" in proj


def main() -> int:
    parser = argparse.ArgumentParser(description="nf-dream translate pre-scan")
    parser.add_argument("scope")
    parser.add_argument("--project")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()

    scope = Path(args.scope).expanduser().resolve()
    if not scope.is_dir():
        print(f"ERROR: {scope} is not a directory", file=sys.stderr)
        return 1

    files: list[dict] = []
    for f in iter_memory_files(scope):
        text = f.read_text()
        parsed = parse_frontmatter(text)
        fm = extract_frontmatter_dict(parsed[0]) if parsed else {}
        if not project_matches(fm, args.project):
            continue
        files.append(scan_file(f))

    summary = {
        "scope": str(scope),
        "project": args.project,
        "total": len(files),
        "by_bucket": {b: sum(1 for f in files if f["bucket"] == b) for b in ["none", "light", "mixed", "heavy"]},
        "candidates": sum(1 for f in files if f["candidate"]),
    }

    if args.json:
        print(json.dumps({"summary": summary, "files": files}, indent=2))
        return 0

    print(f"nf-dream translate pre-scan — {scope}")
    print(f"  total files: {summary['total']}")
    for b, n in summary["by_bucket"].items():
        print(f"  {b}: {n}")
    print(f"  candidates (mixed+heavy): {summary['candidates']}")
    print()
    print("Candidates:")
    for it in [f for f in files if f["candidate"]][:30]:
        print(f"  [{it['bucket']:5s}] {it['file']}  ({it['vn_lines']}/{it['total_lines']} VN lines)")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: nf-dream/scripts/translate-prescan.ts
#!/usr/bin/env bun
/**
 * nf-dream translate pre-scan.
 *
 * For each filtered file, count lines containing non-English diacritics.
 * The shipped detection targets Vietnamese diacritics; edit the DIACRITICS
 * constant below to match your own language.
 *
 * Buckets: none / light / mixed / heavy. Candidates: mixed + heavy.
 *
 * Usage:
 *   bun translate-prescan.ts <scope> [--project <slug>] [--json] [--include-generic]
 */
import { readFileSync, statSync } from "node:fs";
import { resolve, basename } from "node:path";
import { extractFrontmatterDict, iterMemoryFiles, parseFrontmatter } from "./_lib";

const DIACRITICS =
  "àáảãạăắằẳẵặâấầẩẫậ" +
  "èéẻẽẹêếềểễệ" +
  "ìíỉĩị" +
  "òóỏõọôốồổỗộơớờởỡợ" +
  "ùúủũụưứừửữự" +
  "ỳýỷỹỵ" +
  "đ" +
  "ÀÁẢÃẠĂẮẰẲẴẶÂẤẦẨẪẬ" +
  "ÈÉẺẼẸÊẾỀỂỄỆ" +
  "ÌÍỈĨỊ" +
  "ÒÓỎÕỌÔỐỒỔỖỘƠỚỜỞỠỢ" +
  "ÙÚỦŨỤƯỨỪỬỮỰ" +
  "ỲÝỶỸỴ" +
  "Đ";

function escapeForCharClass(s: string): string {
  return s.replace(/[\\\]\-\^]/g, "\\$&");
}

const DIACRITIC_RE = new RegExp(`[${escapeForCharClass(DIACRITICS)}]`);

function bucketOf(vnLines: number, totalLines: number): "none" | "light" | "mixed" | "heavy" {
  if (vnLines === 0) return "none";
  if (vnLines <= 2) return "light";
  const ratio = vnLines / Math.max(totalLines, 1);
  return ratio >= 0.20 ? "heavy" : "mixed";
}

interface ScanResult {
  file: string;
  total_lines: number;
  vn_lines: number;
  bucket: "none" | "light" | "mixed" | "heavy";
  candidate: boolean;
}

function scanFile(path: string): ScanResult {
  const text = readFileSync(path, "utf8");
  const parsed = parseFrontmatter(text);
  const body = parsed ? parsed.body : text;
  const lines = body.split("\n");
  const vnLines = lines.filter((ln) => DIACRITIC_RE.test(ln)).length;
  const bucket = bucketOf(vnLines, lines.length);
  return {
    file: basename(path),
    total_lines: lines.length,
    vn_lines: vnLines,
    bucket,
    candidate: bucket === "mixed" || bucket === "heavy",
  };
}

function projectMatches(
  fmDict: Record<string, string>,
  slug: string | null,
  includeGeneric: boolean
): boolean {
  if (slug === null) return true;
  const proj = fmDict["metadata.project"] ?? "";
  if (proj.includes(slug)) return true;
  if (includeGeneric && proj.includes("generic")) return true;
  return false;
}

interface Args {
  scope: string;
  project: string | null;
  json: boolean;
  includeGeneric: boolean;
  help: boolean;
}

function parseArgs(argv: string[]): Args {
  const a: Args = { scope: "", project: null, json: false, includeGeneric: false, help: false };
  const rest: string[] = [];
  for (let i = 0; i < argv.length; i++) {
    const t = argv[i];
    if (t === "--help" || t === "-h") a.help = true;
    else if (t === "--project") a.project = argv[++i] ?? null;
    else if (t === "--json") a.json = true;
    else if (t === "--include-generic") a.includeGeneric = true;
    else if (t.startsWith("--")) throw new Error(`unknown flag: ${t}`);
    else rest.push(t);
  }
  if (rest.length === 0 && !a.help) throw new Error("missing <scope>");
  a.scope = rest[0] ?? "";
  return a;
}

function printHelp(): void {
  console.log(`bun translate-prescan.ts <scope> [--project <slug>] [--json] [--include-generic]

Pre-scan memory files for non-English content. Reports candidates for translation.`);
}

async function main(): Promise<number> {
  let args: Args;
  try {
    args = parseArgs(process.argv.slice(2));
  } catch (err) {
    console.error(`ERROR: ${(err as Error).message}`);
    printHelp();
    return 1;
  }
  if (args.help) { printHelp(); return 0; }
  const scope = resolve(args.scope.replace(/^~/, process.env.HOME ?? ""));
  try {
    if (!statSync(scope).isDirectory()) throw new Error();
  } catch {
    console.error(`ERROR: ${scope} is not a directory`);
    return 1;
  }

  const files: ScanResult[] = [];
  for (const f of iterMemoryFiles(scope)) {
    const text = readFileSync(f, "utf8");
    const parsed = parseFrontmatter(text);
    const fm = parsed ? extractFrontmatterDict(parsed.fmLines) : {};
    if (!projectMatches(fm, args.project, args.includeGeneric)) continue;
    files.push(scanFile(f));
  }

  const byBucket: Record<string, number> = { none: 0, light: 0, mixed: 0, heavy: 0 };
  for (const f of files) byBucket[f.bucket]++;
  const summary = {
    scope,
    project: args.project,
    total: files.length,
    by_bucket: byBucket,
    candidates: files.filter((f) => f.candidate).length,
  };

  if (args.json) {
    console.log(JSON.stringify({ summary, files }, null, 2));
    return 0;
  }

  console.log(`nf-dream translate pre-scan — ${scope}`);
  console.log(`  total files: ${summary.total}`);
  for (const [b, n] of Object.entries(byBucket)) console.log(`  ${b}: ${n}`);
  console.log(`  candidates (mixed+heavy): ${summary.candidates}`);
  console.log();
  console.log("Candidates:");
  for (const it of files.filter((f) => f.candidate).slice(0, 30)) {
    console.log(`  [${it.bucket.padEnd(5)}] ${it.file}  (${it.vn_lines}/${it.total_lines} non-English lines)`);
  }
  return 0;
}

if (import.meta.main) {
  main().then((code) => process.exit(code));
}
